### ISSUE TITLE
Kokkos Promotion 2.7.00 > 2.7.19

### ISSUE THREAD

**Issue by ndellingwood** [kokkos/kokkos#1845]
kokkos-promotion branch of Trilinos created for testing.

**Comment by ndellingwood** [comment]
Status update:

kokkos develop branch Jenkins jobs clean (aside from `-Werror` with `xl/16.1`), PRs are accounted for.
kokkos-kernels develop branch Jenkins jobs in good state, will merge 2 more PRs then kick off the next round of testing. 

Preliminary testing earlier this week with the kokkos-promotion branch showed that serial (on blake) and openmp (on white) builds were clean with no new errors; threads build on blake had a couple new failures in sacado; cuda needs to be rerun following some fixes in kokkos-kernels and Trilinos that may have impacts. 

Stan reported that there are no issues with LAMMPS with the VOTD Kokkos.

Some updates to the integration testing scripts worth documenting:
- Module devpacks updated to match ATDM testing
- Trilinos updated the minimal supported version of CMake to 3.10.0; currently swapping modules in the scripts after devpacks are loaded to support this
- Issues compiling on White with OpenBLAS require using Netlib version instead; currently swapping modules in the scripts after devpacks are loaded to support this
- Added `m` to `LAPACK_LIBRARIES` to overcome some linking issues, suggested by Si

**Comment by ndellingwood** [comment]
PRs with the last fixes into kokkos-kernels merged; final kokkos round of PR updates; updated kokkos-promotion branch; restarting integration tests.

kokkos develop SHA: 7cd25b3c9fb1c4611fc307662bc2903826a86a03
kokkos-kernels develop SHA: kokkos/kokkos-kernels@50f25eb406efb50ca6a0cbb667563fc933f2f43d
trilinos develop SHA: trilinos/Trilinos@eb2a3fe242a0c8602b87dba35eedfe51a9298ad8
trilinos kokkos-promotion SHA: trilinos/Trilinos@651316ddafab5a8d69488fcc9f69cd78c81814a7

**Comment by ndellingwood** [comment]
Relaunching integrations tests; same SHAs as previous run, but using the kokkos `trilinos-sacado-fix` branch created with PR #1847.

**Comment by ndellingwood** [comment]
Results summary of testing with inclusion of PR #1847:

**Blake**
Serial
- No new build failures
- New failing tests:
```
- Sacado_FadKokkosTests_NoViewSpec_Serial_MPI_1 (Failed)
- Sacado_FadKokkosTests_Serial_MPI_1 (Failed)
```

Threads
- No new build failures
- New failing tests
```
- Sacado_FadKokkosTests_NoViewSpec_Serial_MPI_1 (Failed)
- Sacado_FadKokkosTests_NoViewSpec_Threads_MPI_1 (Failed)
- Sacado_FadKokkosTests_Serial_MPI_1 (Failed)
- Sacado_FadKokkosTests_Threads_MPI_1 (Failed)
```

**White**:
OpenMP
- No new build failures
-New failing tests
Unknown, still waiting in the queue since 10pm last night...

Cuda
- New build failures: Many linking errors of the type 
```
../../../kokkos-kernels/src/libkokkoskernels.so.12.13: undefined reference to `KokkosBlas::Impl::CudaBlasSingleton::singleton()'
collect2: error: ld returned 1 exit status
```
Looked at this with @kyungjoo-kim and it is likely due to missing inclusion of cublas tpl object files in the `src/CMakeLists.txt file`. 
Updated the existing Cuda build but waiting on a Kepler node to retest. Will rebuild/rerun on Pascal node if jobs are taking too long.

**Comment by ndellingwood** [comment]
Fix in kokkos/kokkos-kernels#328 fixed all the linking errors, Cuda builds to 100% now.

**Comment by ndellingwood** [comment]
It looks like the failing Sacado tests that fail are related to shmem size changes:

For example, in the serial build:
```
The following tests FAILED:
    32. Kokkos_View_Fad_SFadType_LayoutLeft_Serial_ShmemSize_UnitTest ... 
    65. Kokkos_View_Fad_SFadType_LayoutRight_Serial_ShmemSize_UnitTest ... 
    98. Kokkos_View_Fad_SFadType_LeftContiguous_Serial_ShmemSize_UnitTest ... 
    131. Kokkos_View_Fad_SFadType_RightContiguous_Serial_ShmemSize_UnitTest ... 
    164. Kokkos_View_Fad_SLFadType_LayoutLeft_Serial_ShmemSize_UnitTest ... 
    196. Kokkos_View_Fad_SLFadType_LayoutRight_Serial_ShmemSize_UnitTest ... 
    228. Kokkos_View_Fad_SLFadType_LeftContiguous_Serial_ShmemSize_UnitTest ... 
    260. Kokkos_View_Fad_SLFadType_RightContiguous_Serial_ShmemSize_UnitTest ...
```

Some specific sample output:
```
32. Kokkos_View_Fad_SFadType_LayoutLeft_Serial_ShmemSize_UnitTest ...
 shmem_size = 3744 == shmem_size_expected = 3696 : FAILED ==> /ascldap/users/ndellin/IntegrationTests-Blake/serial/trilinos-update/packages/sacado/test/UnitTests/Fad_KokkosTests.hpp:1607
 [FAILED]  (9.5e-06 sec) Kokkos_View_Fad_SFadType_LayoutLeft_Serial_ShmemSize_UnitTest
 Location: /ascldap/users/ndellin/IntegrationTests-Blake/serial/trilinos-update/packages/sacado/test/UnitTests/Fad_KokkosTests.hpp:1582
```

**Comment by ndellingwood** [comment]
kokkos-promotion branch of Trilinos updated with changes to Sacado's shmem size unit test checks to match Kokkos updates. I confirmed the previous failing tests on with the serial and threads backends are now passing. Thanks @crtrott for the suggestion!
Will restart testing shortly.

**Comment by ndellingwood** [comment]
Restarting integration tests:

kokkos develop SHA: 21e5f5ba2d603f2420403ec6b126bcbb09f708fe
kokkos-kernels develop SHA: kokkos/kokkos-kernels@8f3004ca44766959b24605acd768653bb5ff08ba
trilinos develop SHA: trilinos/Trilinos@a6353b719597bbcfdaa5c0838b44eeb9b2208488
trilinos kokkos-promotion SHA: trilinos/Trilinos@d4e31c6b3f5f65574ec35f95bec3770ee1f07857

**Comment by ndellingwood** [comment]
Updated testing results:

## Blake
**Serial**

- No new build failures

- New failing tests:
`ROL_adapters_tpetra_test_sol_TpetraSimulatedConstraintInterfaceCVaR_MPI_4`

Failure message:
```
6: [node01:179792] *** An error occurred in MPI_Allreduce
6: [node01:179792] *** reported by process [139666450546689,3]
6: [node01:179792] *** on communicator MPI_COMM_WORLD
6: [node01:179792] *** MPI_ERR_OTHER: known error not in list
6: [node01:179792] *** MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort,
6: [node01:179792] ***    and potentially your MPI job)
6: [node01.blake.sandia.gov:179784] 1 more process has sent help message help-mpi-errors.txt / mpi_errors_are_fatal
6: [node01.blake.sandia.gov:179784] Set MCA parameter "orte_base_help_aggregate" to 0 to see all help / error messages
```

**Threads**

- No new build failures; common ending message of both builds:
`error: node01: task 0: Exited with exit code 2`

- New failing tests
`Same as serial`


## White:
**OpenMP**

- No new build failures
- No new failing tests

**Cuda**

- No new build failures
- New failing tests
```
KokkosCore_UnitTest_Cuda_MPI_1 (Failed)
TpetraCore_MultiVector_UnitTests_MPI_4 (Failed)
```

Failing test output

Kokkos test `KokkosCore_UnitTest_Cuda_MPI_1`
```
[ RUN      ] cuda.cxx11
terminate called after throwing an instance of 'std::runtime_error'
  what():  cudaDeviceSynchronize() error( cudaErrorIllegalAddress): an illegal memory access was encountered /ascldap/users/ndellin/IntegrationTests-White/cuda/trilinos-update/packages/kokkos/core/src/Cuda/Kokkos_Cuda_Impl.cpp:119
Traceback functionality not available
```

Tpetra test `TpetraCore_MultiVector_UnitTests_MPI_4`
```
134:   gblSuccess = 0 == 1 : FAILED ==> /ascldap/users/ndellin/IntegrationTests-White/cuda/trilinos-update/packages/tpetra/core/test/MultiVector/MultiVector_UnitTests.cpp:3968
134:   We modified X_gbl in host memory, and sync'd to device memory, but X_lcl did not change!
134:   gblSuccess = 0 == 1 : FAILED ==> /ascldap/users/ndellin/IntegrationTests-White/cuda/trilinos-update/packages/tpetra/core/test/MultiVector/MultiVector_UnitTests.cpp:3978
134:  [FAILED]  (0.00327 sec) MultiVector_int_int_double_Kokkos_Compat_KokkosCudaWrapperNode_ViewCtor_UnitTest
134:  Location: /ascldap/users/ndellin/IntegrationTests-White/cuda/trilinos-update/packages/tpetra/core/test/MultiVector/MultiVector_UnitTests.cpp:3766
```

**Comment by ndellingwood** [comment]
Kokkos issue #1850 opened regarding the failing Kokkos test; @crtrott has a fix.

**Comment by ndellingwood** [comment]
@crtrott debugged and gave solution for the failing Tpetra tests  - some recent changes in `DualView` led to behavior changes in `get_device_side` that should have been guarded within deprecated code macros. PR #1852 addresses this and fixed the issue in the Cuda Trilinos integration build.

**Comment by ndellingwood** [comment]
Relaunched integration tests.

kokkos dualview-behavior branch (PR #1853) 0eaad7f3c2617ec15470bbfd274571050a1a8da2
kokkos-kernels develop branch: SHA kokkos/kokkos-kernels@66053e268007e6a8bd5e07d7bd2fc9cea46bf207
trilinos develop SHA: trilinos/Trilinos@74384b3a6a7c4b28a72b8bb9d0ee8d8dd04730f4
trilinos kokkos-promotion SHA: trilinos/Trilinos@d4e31c6

**Comment by crtrott** [comment]
I am working on the deprecation stuff.

**Comment by ndellingwood** [comment]
PR #1860 and trilinos/Trilinos#3682 contain changes (on kokkos-promotion branch) to get tpetra tests passing when deprecated code is disabled.

**Comment by ibaned** [comment]
This promotion has been delayed way too long... can we just revert the DualView strictness PR and finish the promotion without it?

**Comment by ibaned** [comment]
We also need to tell LAMMPS people about this... (Stan)

**Comment by stanmoore1** [comment]
We do use the now deprecated `modified_device` and `modified_host` variables directly in a few places in LAMMPS, is that what you are referring to?

**Comment by ndellingwood** [comment]
Reference for testing with deprecated code disabled as of Oct 24:

kokkos: Use branch [issue-1858](https://github.com/kokkos/kokkos/pull/1860)
kokkos-kernels: Use branch [fix-deprecated-code](https://github.com/kokkos/kokkos-kernels/pull/334)
trilinos: Use branch [kokkos-promotion-deprecation-changes](https://github.com/trilinos/Trilinos/pull/3682)

Any changes in Trilinos should be pushed to `kokkos-promotion-deprecation-changes` branch.

Steps for building, running tests on White testbed configured for Cuda_Serial on the Kepler nodes:

1. Clone kokkos, kokkos-kernels, and trilinos to the testbed. Checkout the branches above.
2. In `Trilinos` base directory, create a symbolic links to your `kokkos` and `kokkos-kernels` clones from step 1.
3. Create a `Build` subdirectory within Trilinos' base directory (or do an out-of-source build and adjust the configure script paths accordingly)
4. `cd Build`
5. Copy the attached configure script to the Build directory; adjust paths in the configure script based on your cloned repo location.
6. Source the configure script - this will load appropriate modules and set environment variables.
7. Begin an interactive session on queue with the Kepler GPUs:
`bsub -J Test-DepCode -W 06:00 -Is -n 16 -q rhel7F bash`
8. Build Trilinos
`make -j32`
9. Run tests:
`ctest`
[configure-tpetra.txt](https://github.com/kokkos/kokkos/files/2512065/configure-tpetra.txt)

@ibaned @dhollman @crtrott

**Comment by ibaned** [comment]
@stanmoore1 things are going the other way, @crtrott is moving towards `sync_host` and `sync_device` instead of `sync<SomeHostType>`. I agree with this move, I just think it will have a big impact on LAMMPS/SPARTA.

**Comment by stanmoore1** [comment]
How does that work when the whole kernel is templated on `<DeviceType>` and it could be either host or device?

**Comment by ibaned** [comment]
I'm not sure if the templated versions are going away... this is also less of an issue if you don't use UVM.

**Comment by stanmoore1** [comment]
Tell Christian he's fixing LAMMPS if he breaks it that bad.

**Comment by ibaned** [comment]
We'll probably keep it around I think, this is mostly for Trilinos at this point.

**Comment by crtrott** [comment]
Yeah we need the templated versions. Its just that we also need the non-templated ones if you call the functions inside of non-templated code.

**Comment by ibaned** [comment]
Yep sorry for the scare @stanmoore1 , LAMMPS should be fine

**Comment by mhoemmen** [comment]
> Edit: Building for the Kepler nodes may require including `-D Tpetra_ASSUME_CUDA_AWARE_MPI=OFF` in the configuration script.

You can also control this behavior at run time by setting the environment variable `TPETRA_ASSUME_CUDA_AWARE_MPI` to `0`.

**Comment by ndellingwood** [comment]
Good to know, thanks @mhoemmen !

**Comment by ndellingwood** [comment]
Failing tests with deprecated code disabled in the Tpetra Cuda build are now all fixed. That PR branch is merged into kokkos-promotion branch. 
Waiting on spot-checks for kokkos/kokkos-kernels#334 and kokkos #1860, then ready for retest.

**Comment by ndellingwood** [comment]
Update summary of rerun integration tests (deprecated code enabled):

kokkos develop SHA: c1510cec40f435c92eae84711b1d96a4cc1366cb
kokkos-kernels develop SHA: kokkos/kokkos-kernels@17b0e57586f7a0410a0b63840fbe89b473cde2ae
trilinos develop SHA: trilinos/Trilinos@64a79789e785a4c3066325ebaa91c9b026081712
trilinos kokkos-promotion SHA: trilinos/Trilinos@ed3594facbb3b567485a28857bddc3c55d1ab305
( following merge of trilinos/Trilinos@9dbbc9a )

#### White

**Builds**: No new build failures

**Tests**: The following test now fails with cuda (kepler node) and openmp:
`TpetraCore_MultiVector_UnitTests_MPI_4 (Failed)`

#### Blake 

Builds: 

Tests: The following test now fails with serial and pthread builds (intel compiler)
`TpetraCore_MultiVector_UnitTests_MPI_4`

The failing Tpetra test needs changes that I only applied with deprecated code off, removing the deprecated code guards fixed the issue locally. I'll push that change to kokkos-promotion branch of Trilinos with more fixes needed by other packages with deprecated code disabled.

**Comment by ndellingwood** [comment]
Fix for the failing test `TpetraCore_MultiVector_UnitTests_MPI_4` in kokkos-promotion branch of Trilinos.

Several more fixes pushed to kokkos-promotion branch when deprecated code is disabled. Cuda_Serial build completes 100% with tpetra, sacado, stokhos, muelu, intrepid2, panzer enabled. Some tests fixed, the following two new stokhos tests are failing:

`Stokhos_TpetraCrsMatrixUQPCEUnitTest_Serial_MPI_4 (Failed)`
Error message: 
```
46: Kokkos::HostSpace::allocate[ STD_MALLOC ]( 68719477200 ) FAILED NULL
46: Kokkos::HostSpace::allocate[ STD_MALLOC ]( 68719477200 ) FAILED NULL
```

`Stokhos_TpetraCrsMatrixUQPCEUnitTest_Cuda_MPI_4 (Failed)`
All sub-tests failing within this test, sample error message:
`47:   cudaMallocManaged( &ptr, arg_alloc_size , cudaMemAttachGlobal ) error( cudaErrorInvalidValue): invalid argument ../../kokkos/core/src/Cuda/Kokkos_CudaSpace.cpp:195`

**Comment by ndellingwood** [comment]
@etphipp is looking into the failing stokhos tests, thanks Eric!

**Comment by ndellingwood** [comment]
With deprecated code enabled these Stokhos tests are failing on waterman:

```
	996 - Stokhos_KokkosCrsMatrixUQPCEUnitTest_Cuda_MPI_1 (Failed)
	998 - Stokhos_TpetraCrsMatrixUQPCEUnitTest_Cuda_MPI_4 (Failed)
	1012 - Stokhos_KokkosArrayKernelsUnitTest_Cuda_MPI_1 (Failed)
```

`Stokhos_KokkosCrsMatrixUQPCEUnitTest_Cuda_MPI_1` message:
```
996: 5. Kokkos_CrsMatrix_PCE_UQ_PCE_DS_KokkosMultiply_Multiply_UnitTest ... 
996:  
996:  p=0: *** Caught standard std::exception of type 'std::logic_error' :
996:  
996:   ../../packages/stokhos/src/kokkos/Cuda/Stokhos_Cuda_DeviceProp.hpp:197:
996:   
996:   Throw number = 1
996:   
996:   Throw test that evaluated to true: true
996:   
996:   DeviceProp not configured for compute capability 7
996:  [FAILED]  (0.00355 sec) Kokkos_CrsMatrix_PCE_UQ_PCE_DS_KokkosMultiply_Multiply_UnitTest
996:  Location: ../../packages/stokhos/test/UnitTest/Stokhos_KokkosCrsMatrixUQPCEUnitTest.hpp:704
996:  
996: 
996: The following tests FAILED:
996:     5. Kokkos_CrsMatrix_PCE_UQ_PCE_DS_KokkosMultiply_Multiply_UnitTest ... 
996: 
996: Total Time: 0.0703 sec
996: 
996: Summary: total = 6, run = 6, passed = 5, failed = 1
996: 
996: End Result: TEST FAILED
```


`Stokhos_TpetraCrsMatrixUQPCEUnitTest_Cuda_MPI_4` message:
```
998: 5. Tpetra_CrsMatrix_PCE_DS_int_int_CudaWrapperNode_MatrixVectorMultiply_UnitTest ... 
998:  
998:  p=0: *** Caught standard std::exception of type 'std::logic_error' :
998:  
998:   ../../packages/stokhos/src/kokkos/Cuda/Stokhos_Cuda_DeviceProp.hpp:197:
998:   
998:   Throw number = 1
998:   
998:   Throw test that evaluated to true: true
998:   
998:   DeviceProp not configured for compute capability 7
```

`Stokhos_KokkosArrayKernelsUnitTest_Cuda_MPI_1 ` message:
```
1012: 7. Kokkos_SG_SpMv_double_Cuda_CrsProductTensor_UnitTest ... 
1012:  
1012:  p=0: *** Caught standard std::exception of type 'std::logic_error' :
1012:  
1012:   ../../packages/stokhos/src/kokkos/Cuda/Stokhos_Cuda_DeviceProp.hpp:197:
1012:   
1012:   Throw number = 1
1012:   
1012:   Throw test that evaluated to true: true
1012:   
1012:   DeviceProp not configured for compute capability 7
```

**Comment by ndellingwood** [comment]
Waterman results Cuda_Serial build with deprecated code disabled:

```
Subproject Time Summary:
Ifpack2        = 1394.91 sec*proc (42 tests)
Intrepid2      = 1982.36 sec*proc (255 tests)
MueLu          = 9604.05 sec*proc (99 tests)
Panzer         = 12655.56 sec*proc (157 tests)
Phalanx        = 138.45 sec*proc (27 tests)
Sacado         = 824.47 sec*proc (300 tests)
Stokhos        = 2808.17 sec*proc (84 tests)
Stratimikos    =  74.82 sec*proc (39 tests)
Tpetra         = 4190.44 sec*proc (189 tests)

Total Test time (real) = 2616.70 sec

The following tests FAILED:
	996 - Stokhos_KokkosCrsMatrixUQPCEUnitTest_Cuda_MPI_1 (Failed)
	997 - Stokhos_TpetraCrsMatrixUQPCEUnitTest_Serial_MPI_4 (Failed)
	998 - Stokhos_TpetraCrsMatrixUQPCEUnitTest_Cuda_MPI_4 (Timeout)
	1012 - Stokhos_KokkosArrayKernelsUnitTest_Cuda_MPI_1 (Failed)
```

The stokhos errors are the same as reported with deprecated code enabled.

**Comment by etphipp** [comment]
Progress update:  I tried a OpenMP+Clang build, and all tests passed.  I tried a Cuda+GCC 4.9 build, and got a build error in Tpetra.  I am now trying a Cuda+GCC 7.3 build.  If that doesn't work, I will go to waterman.

**Comment by etphipp** [comment]
For reference, the error I get in Tpetra for a Cuda 9.2 + GCC 4.9 build is:
```
[ 59%] Building CXX object packages/tpetra/core/src/CMakeFiles/tpetra.dir/Tpetra_Experimental_BlockCrsMatrix_DOUBLE_INT_INT_CUDA.cpp.o
/home/etphipp/Trilinos/Trilinos/packages/tpetra/core/src/Tpetra_Experimental_BlockCrsMatrix_def.hpp: In instantiation of void Tpetra::Experimental::BlockCrsMatrix<SC, LO, GO, NT>::setAllToScalar(const Scalar&) [with Scalar = double; LO = int; GO = int; Node = Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Cuda>]:
/home/etphipp/Trilinos/build/opt_mpi_stokhos/packages/tpetra/core/src/Tpetra_Experimental_BlockCrsMatrix_DOUBLE_INT_INT_CUDA.cpp:78:41:   required from here
/home/etphipp/Trilinos/Trilinos/packages/tpetra/core/src/Tpetra_Experimental_BlockCrsMatrix_def.hpp:969:69: error: no matching function for call to Tpetra::Experimental::BlockCrsMatrix<double, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Cuda> >::getValues()
       Kokkos::deep_copy (this->template getValues<device_type> (), alpha);
```

**Comment by ndellingwood** [comment]
Results of tests on waterman using the ATDM scripts:

**deprecated code enabled**:
```
99% tests passed, 3 tests failed out of 1192

Subproject Time Summary:
Ifpack2        = 616.41 sec*proc (42 tests)
Intrepid2      = 660.20 sec*proc (255 tests)
MueLu          = 3850.39 sec*proc (99 tests)
Panzer         = 5151.61 sec*proc (157 tests)
Phalanx        =  63.20 sec*proc (27 tests)
Sacado         = 681.81 sec*proc (300 tests)
Stokhos        = 399.68 sec*proc (84 tests)
Stratimikos    = 110.72 sec*proc (39 tests)
Tpetra         = 2217.18 sec*proc (189 tests)

Total Test time (real) = 5416.42 sec

The following tests FAILED:
	996 - Stokhos_KokkosCrsMatrixUQPCEUnitTest_Cuda_MPI_1 (Failed)
	998 - Stokhos_TpetraCrsMatrixUQPCEUnitTest_Cuda_MPI_4 (Failed)
	1012 - Stokhos_KokkosArrayKernelsUnitTest_Cuda_MPI_1 (Failed)
```

**deprecated code disabled**:
```
99% tests passed, 4 tests failed out of 1284

Subproject Time Summary:
Belos          = 1022.23 sec*proc (92 tests)
Ifpack2        = 605.76 sec*proc (42 tests)
Intrepid2      = 652.46 sec*proc (255 tests)
MueLu          = 3778.71 sec*proc (99 tests)
Panzer         = 4870.67 sec*proc (157 tests)
Phalanx        =  61.03 sec*proc (27 tests)
Sacado         = 665.63 sec*proc (300 tests)
Stokhos        = 2880.02 sec*proc (84 tests)
Stratimikos    = 109.41 sec*proc (39 tests)
Tpetra         = 2127.41 sec*proc (189 tests)

Total Test time (real) = 6167.13 sec

The following tests FAILED:
	1088 - Stokhos_KokkosCrsMatrixUQPCEUnitTest_Cuda_MPI_1 (Failed)
	1089 - Stokhos_TpetraCrsMatrixUQPCEUnitTest_Serial_MPI_4 (Failed)
	1090 - Stokhos_TpetraCrsMatrixUQPCEUnitTest_Cuda_MPI_4 (Timeout)
	1104 - Stokhos_KokkosArrayKernelsUnitTest_Cuda_MPI_1 (Failed)
```

Eric has a fix ready to push for the Stokhos issues, but it brings up a question about a behavior change in the View that popped up in a debug build in Tpetra or Teuchos - in the debug check there it is expected that a View with extent(s) of 0 also has a nullptr, but this is no longer the case if the View was not default constructed (see e.g. #1870)
@crtrott @ibaned

**Comment by etphipp** [comment]
I pushed the stokhos changes that fix the observed failures in the stokhos tests when deprecated code is disabled.

Tpetra and Teuchos are not unit testing `Kokkos::View`.  Rather it is an interoperability issue between Kokkos and Teuchos.  The issue is Tpetra wraps the pointer allocated inside `Kokkos::View` with `Teuchos::ArrayView` when packing/unpacking data for MPI communication (see https://github.com/trilinos/Trilinos/blob/develop/packages/tpetra/core/src/Tpetra_CrsMatrix_def.hpp, line 6704):
```
      auto exportLIDs_h = exportLIDs.template view<HostSpace> ();
      Teuchos::ArrayView<const LO> exportLIDs_av (exportLIDs_h.data (),
                                                  exportLIDs_h.size ());
```
Teuchos expects that if its size argument is 0, the pointer is null (see https://github.com/trilinos/Trilinos/blob/develop/packages/teuchos/core/src/Teuchos_ArrayView.hpp, line, 92):
```
template<class T> inline
ArrayView<const T>::ArrayView(const T* p, size_type size_in, const ERCPNodeLookup rcpNodeLookup )
  : ptr_(p), size_(size_in)
{
#ifdef HAVE_TEUCHOS_ARRAY_BOUNDSCHECK
  TEUCHOS_TEST_FOR_EXCEPT( p != 0 && size_in <= 0 );
```
This causes a unit test in Stokhos to fail in a debug build when deprecated code is disabled.  My guess is similar failures will occur in Tpetra, Ifpack2, and MueLu in debug builds, but I have not tested them.  With deprecated code enabled, the pointer is null and all is good.

I don't have a strong opinion on how this is fixed, either Kokkos needs to revert back to its previous behavior, or the debug check needs to be removed in Teuchos (although I believe there are similar checks in other Teuchos memory management classes that probably should be changed too).  

Personally, I think it is safer to set the pointer to null when the span is 0, since any attempt at accessing it will immediately seg fault, regardless if bounds checking is enabled.

@bartlettroscoe @mhoemmen

**Comment by bartlettroscoe** [comment]
What is the value in having an array view of size zero with a non-null pointer?  That seems like asking for trouble (and hence the Teuchos debug-mode check).

**Comment by ndellingwood** [comment]
Results of tests on waterman using the ATDM scripts - debug build:

deprecated code disabled:
99% tests passed, 5 tests failed out of 1283

Subproject Time Summary:
Belos          = 556.10 sec*proc (91 tests)
Ifpack2        = 913.23 sec*proc (42 tests)
Intrepid2      = 542.29 sec*proc (255 tests)
MueLu          = 4134.95 sec*proc (99 tests)
Panzer         = 5251.84 sec*proc (157 tests)
Phalanx        =  26.82 sec*proc (27 tests)
Sacado         = 204.74 sec*proc (300 tests)
Stokhos        = 2892.05 sec*proc (84 tests)
Stratimikos    =  45.19 sec*proc (39 tests)
Tpetra         = 1633.45 sec*proc (189 tests)

Total Test time (real) = 5592.46 sec

The following tests FAILED:
	312 - TpetraCore_BlockCrsMatrix_MPI_4 (Failed)
	1087 - Stokhos_KokkosCrsMatrixUQPCEUnitTest_Cuda_MPI_1 (Failed)
	1088 - Stokhos_TpetraCrsMatrixUQPCEUnitTest_Serial_MPI_4 (Failed)
	1089 - Stokhos_TpetraCrsMatrixUQPCEUnitTest_Cuda_MPI_4 (Timeout)
	1103 - Stokhos_KokkosArrayKernelsUnitTest_Cuda_MPI_1 (Failed)

**Comment by ibaned** [comment]
The reason for it is that whenever users ask us how to check whether a View has been constructed (even with zero span), our official answer is `data() != nullptr`. Without this change, that recommendation becomes incorrect for zero-span views. Also, for most implementations I've seen, `malloc(0) != nullptr`, so this check seems like more trouble that its worth. I recommend that in the conversion from `Kokkos::View` to `Teuchos::ArrayView`, we simply put in a little if statement to set the pointer to `nullptr` if size is zero to silence Teuchos.

**Comment by etphipp** [comment]
That's a pretty big burden on the user.  Who is going to go through the entire Tpetra stack and find every place this happens and put in the conditional check?

**Comment by ibaned** [comment]
Likely the same people who have been fixing all other breaking changes in each Kokkos release?

**Comment by etphipp** [comment]
So as I understand it, a view that is default constructed has a null pointer, but a non-default constructed view with zero extents has a non-null pointer?  I am curious, what is the use case where someone would need to distinguish between those two cases?

**Comment by bartlettroscoe** [comment]
> In my own code a I have a couple kernels where some of the input Views are optional, and one can simply check inside the kernel whether `data() == nullptr` to decide whether or not to do extra computation based on numbers in those Views.

@ibaned, but in that case if the view has `size()==0`, you would hot do any computation based on those numbers either, do you?  If an array view can return `data() != nullptr` when `size() == 0`, then would that not mess up your algorithm or require it to do work it may not have to do can call functions it should not have to call?

**Comment by mhoemmen** [comment]
@etphipp Wait, hold on, can't we just fix that "wrap Kokkos::View in Teuchos::ArrayView" compatibility wrapper thing?

**Comment by mhoemmen** [comment]
That should fix a lot of the instances, no?

**Comment by etphipp** [comment]
Well if the code used it, then yes.  But it doesn't.

**Comment by mhoemmen** [comment]
:( ugh that sucks :( had thought we had fixed this in Teuchos a while back to permit non-null zero-length things.

**Comment by ibaned** [comment]
@bartlettroscoe its not that simple, the mere presence (allocated status) of the View carries information and alters the outcome, even if it is zero length.

I'd much rather remove the Teuchos check or do the necessary refactor in Tpetra to use a single wrapper thus reducing code duplication instead of keeping a bug and losing debug capability in Kokkos just to workaround unnecessary checks in Teuchos and copy-paste programming in Tpetra.

**Comment by etphipp** [comment]
It does feel like we've run into this before.  Maybe we changed ArrayRCP or something?

Again, I don't have strong opinions on this.  My only point was that it is safer for the user to have the pointer be null when the size is zero, since it gives rise to immediate seg faults which are much easier to debug.  But if Kokkos has a valid reason for doing this, then just take the check out of Teuchos.

**Comment by bartlettroscoe** [comment]
@ibaned, does `Kokkos::View:<...>::data()` guaranteed that any non-null pointer returned points to valid memory that can be accessed? 

Can we make it so that `Teuchos::ArrayView` stores such a pointer provides a new function (with some logical name) to return this special data pointer but still have `Teuchos::ArrayView::get()` and `Teuchos::ArrayView::getRawPtr()` still return `nullptr`?  And then the converters to and from `Teuchos::ArrayView` and `Kokkos::View` can work with this special pointer?  That would preserve backward compatibility for code that uses `Teuchos::ArrayView` and would give `Kokkos::View` what is wants.

**Comment by ibaned** [comment]
Nothing other than non-nullness is guaranteed in the zero-span case.  Guarantees in the non-zero span case depend on the layout etc.

That seems like a lot of Kokkos-specific complexity in ArrayView. Is anyone in Trilinos really relying on this equivalence between `getRawPtr()==nullptr` and `size()==0`?

**Comment by bartlettroscoe** [comment]
> Nothing other than non-nullness is guaranteed in the zero-span case. Guarantees in the non-zero span case depend on the layout etc.

It is not a good idea to expose a raw pointer in my opinion.  Instead, can you add a `Kokkos::View::isinitialized()` function to see if a view as been initialized (in addition to perhaps zero range)? 

> That seems like a lot of Kokkos-specific complexity in ArrayView.

It would be non-ideal but not the end of the world.  But it would be much better than having `Teuchos::ArrayView::getRawPtr()` return non-null when `size()==0`.

>  Is anyone in Trilinos really relying on this equivalence between `getRawPtr()==nullptr and size()==0`?

I can almost guarantee the answer is yet.  In fact, I think I have written a lot of code like this.  In fact, that is why I added `arrayView.getRawPtr()` instead of using `&arrayView[0]` because the latter throws when the view is unsized.

**Comment by ndellingwood** [comment]
@etphipp was the error [here](https://github.com/kokkos/kokkos/issues/1845#issuecomment-435116457) with a non-MPI build? I reproduced with cuda/9.2 + gcc/4.9.3 in a non-MPI build, all the other tests I had checked with MPI. If you're build was non-MPI it could explain why it didn't happen in my builds.

**Comment by ndellingwood** [comment]
@crtrott 
SPARC built 100% on Waterman with the following:

kokkos develop branch SHA: c1510cec40f435c92eae84711b1d96a4cc1366cb
kokkos-kernels develop branch SHA: kokkos/kokkos-kernels@79a862434a3804062e471d7842f4220b2918aca8
trilinos kokkos-promotion branch SHA: trilinos/Trilinos@0aad160feb33c40934e90021ed3becdac928c671

I had to modify the sparc waterman script in the repo to get Trilinos configured/built, can email to interested parties (thanks Eric for some suggestions on sticking points!).

Adding @mhoemmen @micahahoward

**Comment by ndellingwood** [comment]
Fix pushed to kokkos-promotion branch of Trilinos for the failing Tpetra test on Waterman with deprecated code off and debugging on.

**Comment by etphipp** [comment]
> @etphipp was the error here with a non-MPI build? I reproduced with cuda/9.2 + gcc/4.9.3 in a non-MPI build, all the other tests I had checked with MPI. If you're build was non-MPI it could explain why it didn't happen in my builds.

@ndellingwood, no that build had MPI enabled.  But it sounds like you reproduced it anyway and fixed it?

**Comment by mhoemmen** [comment]
I'm not sure if SPARC has tests that exercise a lot of MueLu, but the plan is to use more of that, I think.  MueLu is likely to have a lot of code that uses Teuchos::ArrayView on things that could be UVM allocations.

I know we had to start with UVM to get things working in Trilinos, but I really wish we had had the resources at the time to push forward, finish the conversion, and get rid of the UVM requirement.

**Comment by crtrott** [comment]
I reran Blake and White tests using the scripts in the kokkos-kernels/scripts/trilinos-integration repos. Has some issues with jobs stopping to run but by manually rebuilding it looked all good. Here is test output:

# Blake Pthread 
### Updated:
```
Test project /ascldap/users/crtrott/TrilinosBuild/Blake/TrilinosIntegration/Pthreads/build_updated
        Start   1: KokkosCore_UnitTest_Serial_MPI_1
  1/1764 Test   #1: KokkosCore_UnitTest_Serial_MPI_1 ..............................................................   Passed    4.47 sec
...
        Start 1764: PanzerAdaptersIOSS_tIOSSConnManager3_MPI_3
1764/1764 Test #1764: PanzerAdaptersIOSS_tIOSSConnManager3_MPI_3 ....................................................   Passed    0.90 sec


99% tests passed, 6 tests failed out of 1764

Label Time Summary:
Amesos2          =   4.34 sec (8 tests)
Anasazi          =  51.06 sec (74 tests)
Belos            =  62.71 sec (92 tests)
Ifpack           =  26.64 sec (41 tests)
Ifpack2          =  33.87 sec (42 tests)
Intrepid         = 115.91 sec (144 tests)
Intrepid2        =  46.88 sec (148 tests)
Isorropia        =   2.35 sec (4 tests)
Kokkos           =  80.50 sec (26 tests)
KokkosKernels    = 286.83 sec (8 tests)
ML               =  11.90 sec (16 tests)
MueLu            = 122.14 sec (59 tests)
NOX              =  71.74 sec (92 tests)
Panzer           =  95.98 sec (120 tests)
Phalanx          =  11.50 sec (22 tests)
ROL              = 112.93 sec (84 tests)
Sacado           =  61.26 sec (297 tests)
Stokhos          =  65.03 sec (79 tests)
Stratimikos      =  17.86 sec (23 tests)
Thyra            =  40.19 sec (81 tests)
Tpetra           = 106.56 sec (192 tests)
Xpetra           =   6.92 sec (17 tests)
Zoltan2          =  68.63 sec (101 tests)

Total Test time (real) = 1517.02 sec

The following tests FAILED:
	613 - Xpetra_CrsMatrix_UnitTests_MPI_4 (Not Run)
	615 - Xpetra_Matrix_UnitTests_MPI_4 (Not Run)
	616 - Xpetra_MatrixMatrix_UnitTests_MPI_4 (Not Run)
	618 - Xpetra_BlockedCrsMatrix_UnitTests_MPI_4 (Not Run)
	619 - Xpetra_ThyraBlockedOperator_UnitTests_MPI_4 (Not Run)
	1006 - Stratimikos_test_single_belos_thyra_solver_driver_nos1_nrhs8_MPI_1 (Failed)
Errors while running CTest
```

### Pristine:
```
Test project /ascldap/users/crtrott/TrilinosBuild/Blake/TrilinosIntegration/Pthreads/build_pristine
        Start   1: KokkosCore_UnitTest_Serial_MPI_1
  1/1764 Test   #1: KokkosCore_UnitTest_Serial_MPI_1 ..............................................................   Passed    4.84 sec
        Start   2: KokkosCore_UnitTest_Threads_MPI_1
...
        Start 1764: PanzerAdaptersIOSS_tIOSSConnManager3_MPI_3
1764/1764 Test #1764: PanzerAdaptersIOSS_tIOSSConnManager3_MPI_3 ....................................................   Passed    1.07 sec

99% tests passed, 8 tests failed out of 1764

Subproject Time Summary:
Amesos2          =  21.20 sec*proc (8 tests)
Anasazi          = 391.79 sec*proc (74 tests)
Belos            = 1406.35 sec*proc (92 tests)
Ifpack           = 104.37 sec*proc (36 tests)
Ifpack2          =  97.30 sec*proc (42 tests)
Intrepid         = 870.84 sec*proc (144 tests)
Intrepid2        = 475.42 sec*proc (148 tests)
Isorropia        =  12.64 sec*proc (4 tests)
Kokkos           =  97.84 sec*proc (26 tests)
KokkosKernels    = 210.50 sec*proc (8 tests)
ML               = 113.60 sec*proc (16 tests)
MueLu            = 2190.79 sec*proc (58 tests)
NOX              = 1072.38 sec*proc (92 tests)
Panzer           = 505.14 sec*proc (120 tests)
Phalanx          =  16.14 sec*proc (22 tests)
ROL              = 591.47 sec*proc (84 tests)
Sacado           = 989.74 sec*proc (297 tests)
Stokhos          = 172.88 sec*proc (79 tests)
Stratimikos      =  30.10 sec*proc (23 tests)
Thyra            = 164.78 sec*proc (81 tests)
Tpetra           = 2023.20 sec*proc (192 tests)
Xpetra           = 158.50 sec*proc (17 tests)
Zoltan2          = 1476.67 sec*proc (101 tests)

Total Test time (real) = 6808.63 sec

The following tests FAILED:
	613 - Xpetra_CrsMatrix_UnitTests_MPI_4 (Not Run)
	615 - Xpetra_Matrix_UnitTests_MPI_4 (Not Run)
	616 - Xpetra_MatrixMatrix_UnitTests_MPI_4 (Not Run)
	618 - Xpetra_BlockedCrsMatrix_UnitTests_MPI_4 (Not Run)
	619 - Xpetra_ThyraBlockedOperator_UnitTests_MPI_4 (Not Run)
	1006 - Stratimikos_test_single_belos_thyra_solver_driver_nos1_nrhs8_MPI_1 (Failed)
	1047 - Intrepid_test_Discretization_Basis_HGRAD_LINE_Hermite_FEM_Test_01_MPI_1 (Timeout)
	1579 - ROL_adapters_tpetra_test_sol_TpetraSimulatedConstraintInterfaceCVaR_MPI_4 (Failed)
Errors while running CTest
```

# Blake Serial
### Updated
```
Test project /ascldap/users/crtrott/TrilinosBuild/Blake/TrilinosIntegration/Serial/build_updated
        Start   1: KokkosCore_UnitTest_Serial_MPI_1
  1/1797 Test   #1: KokkosCore_UnitTest_Serial_MPI_1 ..............................................................   Passed    4.64 sec
        Start   2: KokkosCore_UnitTest_Default_MPI_1
...
        Start 1797: PanzerAdaptersIOSS_tIOSSConnManager3_MPI_3
1797/1797 Test #1797: PanzerAdaptersIOSS_tIOSSConnManager3_MPI_3 ....................................................   Passed    0.97 sec

99% tests passed, 2 tests failed out of 1797

Label Time Summary:
Amesos2          =   4.92 sec (9 tests)
Anasazi          =  69.79 sec (101 tests)
Belos            =  74.92 sec (107 tests)
Ifpack           =  26.79 sec (41 tests)
Ifpack2          =  36.60 sec (42 tests)
Intrepid         = 115.63 sec (144 tests)
Intrepid2        =  45.79 sec (148 tests)
Isorropia        =   2.89 sec (4 tests)
Kokkos           =  28.04 sec (24 tests)
KokkosKernels    = 111.04 sec (4 tests)
ML               =  12.04 sec (16 tests)
MueLu            = 132.55 sec (58 tests)
NOX              =  73.83 sec (92 tests)
Panzer           =  99.85 sec (120 tests)
Phalanx          =  11.36 sec (22 tests)
ROL              = 101.47 sec (84 tests)
Sacado           =  58.80 sec (293 tests)
ShyLU_Node       =   0.57 sec (3 tests)
Stokhos          =  66.45 sec (76 tests)
Stratimikos      =  17.99 sec (23 tests)
Thyra            =  40.41 sec (81 tests)
Tpetra           = 112.73 sec (192 tests)
Xpetra           =  13.25 sec (18 tests)
Zoltan2          =  68.72 sec (101 tests)

Total Test time (real) = 1340.94 sec

The following tests FAILED:
	1043 - Stratimikos_test_single_belos_thyra_solver_driver_nos1_nrhs8_MPI_1 (Failed)
	1612 - ROL_adapters_tpetra_test_sol_TpetraSimulatedConstraintInterfaceCVaR_MPI_4 (Failed)
```

### Pristine
```
Test project /ascldap/users/crtrott/TrilinosBuild/Blake/TrilinosIntegration/Serial/build_pristine
        Start   1: KokkosCore_UnitTest_Serial_MPI_1
  1/1797 Test   #1: KokkosCore_UnitTest_Serial_MPI_1 ..............................................................   Passed    3.65 sec
        Start   2: KokkosCore_UnitTest_Default_MPI_1
...
        Start 1797: PanzerAdaptersIOSS_tIOSSConnManager3_MPI_3
1797/1797 Test #1797: PanzerAdaptersIOSS_tIOSSConnManager3_MPI_3 ....................................................   Passed    0.68 sec

99% tests passed, 2 tests failed out of 1797

Label Time Summary:
Amesos2          =   3.03 sec (9 tests)
Anasazi          =  46.90 sec (101 tests)
Belos            =  53.00 sec (107 tests)
Ifpack           =  17.38 sec (41 tests)
Ifpack2          =  27.56 sec (42 tests)
Intrepid         =  74.30 sec (144 tests)
Intrepid2        =  36.51 sec (148 tests)
Isorropia        =   1.54 sec (4 tests)
Kokkos           =  23.31 sec (24 tests)
KokkosKernels    =  88.78 sec (4 tests)
ML               =   8.52 sec (16 tests)
MueLu            = 121.26 sec (58 tests)
NOX              =  51.12 sec (92 tests)
Panzer           =  69.52 sec (120 tests)
Phalanx          =   7.16 sec (22 tests)
ROL              =  76.23 sec (84 tests)
Sacado           =  45.01 sec (293 tests)
ShyLU_Node       =   0.48 sec (3 tests)
Stokhos          =  48.11 sec (76 tests)
Stratimikos      =  12.72 sec (23 tests)
Thyra            =  24.56 sec (81 tests)
Tpetra           =  70.48 sec (192 tests)
Xpetra           =   9.11 sec (18 tests)
Zoltan2          =  48.66 sec (101 tests)

Total Test time (real) = 977.32 sec

The following tests FAILED:
	1043 - Stratimikos_test_single_belos_thyra_solver_driver_nos1_nrhs8_MPI_1 (Failed)
	1612 - ROL_adapters_tpetra_test_sol_TpetraSimulatedConstraintInterfaceCVaR_MPI_4 (Failed)
Errors while running CTest
```
